tests/test_nhg.py

```diff
@@ -384,20 +799,891 @@ def asic_route_nhg_fvs(k):
+        nhg_ps.set("group2", fvs)
+
+        # The group should be created.  To test this, add a route pointing to
+        # it.  If the group exists, the route will be created as well.
+        fvs = swsscommon.FieldValuePairs([("nexthop_group", "group2")])
+        rt_ps.set('2.2.4.0/24', fvs)
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 2)
+
+        # Delete the route and the group
+        rt_ps._del('2.2.4.0/24')
+        nhg_ps._del('group2')
+
+        # Bring the interface back up
+        self.flap_intf(1, 'up', dvs)
+
+        # Create group2 pointing to a NH that does not exist
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.61'), ("ifname", "Ethernet120")])
+        nhg_ps.set("group2", fvs)
+
+        # The group should fail to be created.  To test this, we add a route
+        # pointing to it.  The route should not be created.
+        fvs = swsscommon.FieldValuePairs([("nexthop_group", "group2")])
+        rt_ps.set('2.2.4.0/24', fvs)
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 1)
+
+        # Configure the NH's interface
+        self.config_intf(30, dvs)
+
+        # A couple of more routes will be added to ASIC_DB
+        asic_routes_count += 2
+
+        # The group should be created, so the route should be added
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count + 2)
+
+        # Delete the route and the group
+        rt_ps._del('2.2.4.0/24')
+        nhg_ps._del('group2')
+
+        # Update the group to a multiple NH group - should fail as the group
+        # is referenced
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.1,10.0.0.3'),
+                                          ("ifname", "Ethernet0,Ethernet4")])
+        nhg_ps.set("group1", fvs)
+        time.sleep(1)
+        assert len(asic_db.get_keys(self.ASIC_NHG_STR)) == asic_nhgs_count
+
+        # Update group1 to point to another NH - should fail as the group is
+        # referenced
+        nhgid = self.get_nhg_id('group1', dvs)
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.3'), ("ifname", "Ethernet4")])
+        nhg_ps.set("group1", fvs)
+        assert nhgid == self.get_nhg_id('group1', dvs)
+
+        # Remove route 2.2.2.0/24
+        rt_ps._del("2.2.2.0/24")
+
+        # Wait for route 2.2.2.0/24 to be removed
+        asic_db.wait_for_n_keys(self.ASIC_RT_STR, asic_routes_count)
+
+        # The group is not referenced anymore, so it should be updated
+        assert nhgid != self.get_nhg_id('group1', dvs)
+
+        # Update the group to a multiple NH group - should work as the group is
+        # not referenced
+        fvs = swsscommon.FieldValuePairs([('nexthop', '10.0.0.1,10.0.0.3'),
+                                          ("ifname", "Ethernet0,Ethernet4")])
+        nhg_ps.set("group1", fvs)
+        asic_db.wait_for_n_keys(self.ASIC_NHG_STR, asic_nhgs_count + 1)
+
```

Comment:
Can you also add the reverse test case. i.e. add a NHOP group that's being referenced by a route, update it with single next hop, and check it doesn't get updated.

Both these tests will be useful if and when a solution is in place to handle this case.

Reply:
Done.